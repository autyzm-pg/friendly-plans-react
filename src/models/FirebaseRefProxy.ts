import firebase, { RNFirebase } from 'react-native-firebase';

export const getPlanSubItemRef = (
  studentId: string,
  planId: string,
  planItemId: string,
  planSubItemId: string,
): RNFirebase.firestore.DocumentReference => getPlanSubItemsRef(studentId, planId, planItemId).doc(planSubItemId);

export const getPlanSubItemsRef = (
  studentId: string,
  planId: string,
  planItemId: string,
): RNFirebase.firestore.CollectionReference => getPlanItemRef(studentId, planId, planItemId).collection('subItems');

export const getPlanItemRef = (
  studentId: string,
  planId: string,
  planItemId: string,
): RNFirebase.firestore.DocumentReference => getPlanItemsRef(studentId, planId).doc(planItemId);

export const getPlanItemsRef = (studentId: string, planId: string): RNFirebase.firestore.CollectionReference =>
  getPlanRef(studentId, planId).collection('planItems');

export const getPlanRef = (studentId: string, planId: string): RNFirebase.firestore.DocumentReference =>
  getPlansRef(studentId).doc(planId);

export const getPlansRef = (studentId: string): RNFirebase.firestore.CollectionReference =>
  getStudentRef(studentId).collection('plans');

export const getStudentRef = (studentId: string): RNFirebase.firestore.DocumentReference =>
  getStudentsRef().doc(studentId);

export const getStudentsRef = (userId: string = getAuthenticatedUserId()): RNFirebase.firestore.CollectionReference =>
  getUserRef(userId).collection('students');

export const getUserRef = (userId: string): RNFirebase.firestore.DocumentReference => getUsersRef().doc(userId);

const getUsersRef = (): RNFirebase.firestore.CollectionReference => {
  return firebase.firestore().collection('users');
};

export const getAuthenticatedUserId = (): string => firebase.auth().currentUser!.uid;
