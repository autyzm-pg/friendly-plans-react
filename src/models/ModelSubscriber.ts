import { RNFirebase } from 'react-native-firebase';
import { OperationalError } from '../infrastructure/Errors';
import { SubscribableModel } from './SubscribableModel';

export class ModelSubscriber<T> {
  private unsubscribeCollectionUpdatesCallback: any = null;
  private unsubscribeElementUpdatesCallback: any = null;

  subscribeCollectionUpdates = (parent: SubscribableModel, callback: (elements: T[]) => void): void => {
    if (this.unsubscribeCollectionUpdatesCallback !== null) {
      throw new OperationalError(
        'Trying to subscribe on collection updates when there is a subscription already. Unsubscribe first.',
      );
    }
    this.unsubscribeCollectionUpdatesCallback = parent
      .getChildCollectionRef()
      .onSnapshot(this.parseCollectionDataToObjects(parent, callback));
  };

  unsubscribeCollectionUpdates = () => {
    if (this.unsubscribeCollectionUpdatesCallback === null) {
      throw new OperationalError(
        'Trying to unsubscribe on collection updates when there is no subscription. Subscribe first.',
      );
    }
    this.unsubscribeCollectionUpdatesCallback();
    this.unsubscribeCollectionUpdatesCallback = null;
  };

  subscribeElementUpdates = (element: SubscribableModel, callback: (element: T) => void): void => {
    if (this.unsubscribeElementUpdatesCallback !== null) {
      throw new OperationalError(
        'Trying to subscribe on element updates when there is a subscription already. Unsubscribe first.',
      );
    }
    this.unsubscribeElementUpdatesCallback = element
      .getRef()
      .onSnapshot(this.parseSingleElementDataToObject(element, callback));
  };

  unsubscribeElementUpdates = () => {
    if (this.unsubscribeElementUpdatesCallback === null) {
      throw new OperationalError(
        'Trying to unsubscribe on element updates when there is no subscription. Subscribe first.',
      );
    }
    this.unsubscribeElementUpdatesCallback();
    this.unsubscribeElementUpdatesCallback = null;
  };

  private parseCollectionDataToObjects = (
    parent: SubscribableModel,
    callback: (elements: T[]) => void,
  ): ((querySnapshot: RNFirebase.firestore.QuerySnapshot) => void) => {
    return querySnapshot => {
      const ChildType = parent.getChildType();
      const elements: T[] = querySnapshot.docs.map(document => {
        return Object.assign(new ChildType(), {
          id: document.id,
          ...document.data(),
        });
      });
      callback(elements);
    };
  };

  private parseSingleElementDataToObject = (
    element: SubscribableModel,
    callback: (element: T) => void,
  ): ((document: RNFirebase.firestore.DocumentSnapshot) => void) => {
    return document => {
      if (document.exists) {
        const updatedElement = Object.assign(new element.constructor(), {
          id: document.id,
          ...document.data(),
        });
        callback(updatedElement);
      }
    };
  };
}
