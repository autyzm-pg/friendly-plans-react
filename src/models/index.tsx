export * from './DialogProps';
export * from './User';
export * from './Student';
export * from './Plan';
export * from './PlanItem';
export * from './PlanSubItem';
