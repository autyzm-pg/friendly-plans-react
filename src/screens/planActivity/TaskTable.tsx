import React, { SFC } from 'react';
import { StyleSheet, View } from 'react-native';

import { FullScreenTemplate } from 'components';
import { dimensions, getElevation, palette } from 'styles';
import { TableRow } from './TableRow';

export const CELL_DIMENSIONS = {
  NUMBER: 0.3,
  CHECKBOX: 0.3,
  NAME: 1.7,
  TYPE: 0.5,
  SECTION: 0.5,
  TIME: 1.9,
  DELETE: 0.4,
  EDIT: 0.6,
};

interface Props {
  rowList: number[];
}

export const TaskTable: SFC<Props> = ({ rowList }) => (
  <FullScreenTemplate darkBackground>
    <View style={styles.container}>
      {rowList.map((rowNumber, index) => (
        <TableRow border={index !== rowList.length - 1} key={index} rowNumber={rowNumber + 1} />
      ))}
    </View>
  </FullScreenTemplate>
);

const styles = StyleSheet.create({
  container: {
    ...getElevation(1),
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: palette.background,
    marginHorizontal: '6%',
    borderBottomLeftRadius: dimensions.spacingMedium,
    borderBottomRightRadius: dimensions.spacingMedium,
  },
});
