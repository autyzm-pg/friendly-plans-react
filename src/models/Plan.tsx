import { RNFirebase } from 'react-native-firebase';

import { i18n } from 'locale';
import { PlanItemType } from './PlanItem';
import { getStudentsRef } from './Student';

export class Plan {
  name!: string;
  id!: string;
  studentId!: string;

  delete = (): Promise<void> => deletePlan(this);

  update = (changes: object): Promise<void> => updatePlan(this, changes);

  createPlanItem = (
    planItemType: PlanItemType,
  ): Promise<RNFirebase.firestore.DocumentReference> =>
    createPlanItem(this, planItemType);

  getPlanItemsRef = (): RNFirebase.firestore.CollectionReference =>
    getPlanItemsRef(this); 
}

// Private API below
const getPlanRef = (plan: Plan): RNFirebase.firestore.DocumentReference =>
  getStudentsRef()
    .doc(plan.studentId)
    .collection('plans')
    .doc(plan.id);

const deletePlan = (plan: Plan): Promise<void> => getPlanRef(plan).delete();

const updatePlan = (plan: Plan, changes: object): Promise<void> =>
  getPlanRef(plan).update(changes);

const getPlanItemsRef = (
  plan: Plan,
): RNFirebase.firestore.CollectionReference =>
  getPlanRef(plan).collection('planItems');

const createPlanItem = (
  plan: Plan,
  type: PlanItemType,
): Promise<RNFirebase.firestore.DocumentReference> =>
  getPlanItemsRef(plan).add({
    name: i18n.t('updatePlan:planItemNamePlaceholder'),
    studentId: plan.studentId,
    planId: plan.id,
    type,
    completed: false
  });