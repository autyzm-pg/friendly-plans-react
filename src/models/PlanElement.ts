import { PlanItemType } from './PlanItem';

export interface PlanElement {
  id: string;
  name: string;
  type: PlanItemType;
  completed: boolean;
  time: number;
  lector: boolean;

  complete: () => void;
  update: (changes: any) => void;
}
