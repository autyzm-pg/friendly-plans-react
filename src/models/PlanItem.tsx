import { RNFirebase } from 'react-native-firebase';
import { i18n } from '../locale';
import { getPlanItemRef, getPlanItemsRef, getPlanSubItemsRef } from './FirebaseRefProxy';
import { Plan } from './Plan';
import { PlanElement } from './PlanElement';
import { PlanSubItem } from './PlanSubItem';
import { ParameterlessConstructor, SubscribableModel } from './SubscribableModel';

export enum PlanItemType {
  SimpleTask = 'simpleTask',
  ComplexTask = 'complexTask',
  Break = 'break',
  Interaction = 'interaction',
  SubElement = 'subElement',
}

const PLAN_ITEMS_ICONS = {
  task: 'layers',
  break: 'bell',
  interaction: 'account-multiple',
};

export class PlanItem implements SubscribableModel, PlanElement {
  static create = (
    plan: Plan,
    type: PlanItemType,
    name: string = i18n.t('updatePlan:planItemNamePlaceholder'),
  ): Promise<RNFirebase.firestore.DocumentReference> =>
    getPlanItemsRef(plan.studentId, plan.id).add({
      name,
      studentId: plan.studentId,
      planId: plan.id,
      type,
      completed: false,
      lector: false,
      nameForChild: i18n.t('planItemActivity:taskNameForChild'),
    });

  static async createPlanItem(
    plan: Plan,
    type: PlanItemType,
    name: string = i18n.t('updatePlan:planItemNamePlaceholder'),
  ): Promise<PlanItem> {
    const { id } = await getPlanItemsRef(plan.studentId, plan.id).add({
      name,
      studentId: plan.studentId,
      planId: plan.id,
      type,
      completed: false,
      lector: false,
      nameForChild: i18n.t('planItemActivity:taskNameForChild'),
    });

    return Object.assign(new PlanItem(), {
      id,
      name,
      studentId: plan.studentId,
      planId: plan.id,
      type,
      completed: false,
      lector: false,
      nameForChild: i18n.t('planItemActivity:taskNameForChild'),
    });
  }

  id!: string;
  name!: string;
  type!: PlanItemType;
  planId!: string;
  studentId!: string;
  completed!: boolean;
  time!: number;
  image!: string;
  lector!: boolean;
  nameForChild!: string;

  getIconName = (): string => {
    return PLAN_ITEMS_ICONS[this.type];
  };

  isTask = (): boolean => this.type === PlanItemType.SimpleTask || this.type === PlanItemType.ComplexTask;
  isSimpleTask = (): boolean => this.type === PlanItemType.SimpleTask;
  complete = () => {
    this.update({ completed: true });
  };

  changeType = (type: PlanItemType) => {
    this.update({
      type,
    });
  };

  update = (changes: object) => getPlanItemRef(this.studentId, this.planId, this.id).update(changes);
  delete = (): Promise<void> => getPlanItemRef(this.studentId, this.planId, this.id).delete();

  getChildCollectionRef: () => RNFirebase.firestore.CollectionReference = () =>
    getPlanSubItemsRef(this.studentId, this.planId, this.id);
  getChildType: () => ParameterlessConstructor<SubscribableModel> = () => PlanSubItem;
  getRef: () => RNFirebase.firestore.DocumentReference = () => getPlanItemRef(this.studentId, this.planId, this.id);
}
