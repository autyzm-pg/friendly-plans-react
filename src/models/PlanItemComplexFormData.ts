import { PlanItemType } from './PlanItem';

export interface PlanItemComplexFormData {
  name: string;
  time: number;
  taskType: PlanItemType;
}
