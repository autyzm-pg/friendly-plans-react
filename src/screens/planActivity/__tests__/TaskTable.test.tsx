import { shallow } from 'enzyme';
import React from 'react';

import { PlanItem } from 'models';
import { TaskTable } from '../TaskTable';

const props = {
  planItemList: [new PlanItem()],
  handlePlanListOrderChanged: jest.fn(),
};

describe('TaskTable', () => {
  it('renders correctly', () => {
    const taskTable = shallow(<TaskTable {...props} />);

    expect(taskTable).toMatchSnapshot();
  });
});
