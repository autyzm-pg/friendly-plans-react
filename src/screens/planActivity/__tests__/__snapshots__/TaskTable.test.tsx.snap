// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskTable renders correctly 1`] = `
<FullScreenTemplate
  darkBackground={true}
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#FFFFFF",
        "borderBottomLeftRadius": 16,
        "borderBottomRightRadius": 16,
        "elevation": 1,
        "justifyContent": "center",
        "marginHorizontal": "6%",
        "shadowColor": "rgba(0, 0, 0, 0.7)",
        "shadowOffset": Object {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.18000000000000002,
        "shadowRadius": 1.5,
      }
    }
  >
    <Component
      border={false}
      key="0"
      planItem={
        PlanItem {
          "changeType": [Function],
          "complete": [Function],
          "delete": [Function],
          "getChildCollectionRef": [Function],
          "getChildType": [Function],
          "getIconName": [Function],
          "getRef": [Function],
          "isSimpleTask": [Function],
          "isTask": [Function],
          "setComplete": [Function],
          "update": [Function],
        }
      }
      rowNumber={1}
    />
  </View>
</FullScreenTemplate>
`;
